### PR TITLE
[refactor]use clear concept to replace std::enable_if_t

### PR DESCRIPTION
**Pull request by ucasfl**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ucasfl** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.08 seconds
 stream load tsv:          515 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162091070 Bytes

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/vec/core/decimal_comparison.h [hunk 1]
```diff
@@ -135,8 +135,8 @@ class DecimalComparison {
     }
 
     template <typename T, typename U>
-    static std::enable_if_t<IsDecimalNumber<T> && IsDecimalNumber<U>, Shift> getScales(
-            const DataTypePtr& left_type, const DataTypePtr& right_type) {
+        requires IsDecimalNumber<T> && IsDecimalNumber<U>
```
Comment (yiguolei) [inline be/src/vec/core/decimal_comparison.h:138]: requires (IsDecimalNumber<T> && IsDecimalNumber<U>) ?

File: be/src/vec/data_types/data_type_decimal.h [hunk 2]
```diff
@@ -407,10 +407,10 @@ constexpr bool IsDataTypeDecimalOrNumber =
         IsDataTypeDecimal<DataType> || IsDataTypeNumber<DataType>;
 
 template <typename FromDataType, typename ToDataType>
-std::enable_if_t<IsDataTypeDecimal<FromDataType> && IsDataTypeDecimal<ToDataType>,
-                 typename ToDataType::FieldType>
-convert_decimals(const typename FromDataType::FieldType& value, UInt32 scale_from, UInt32 scale_to,
-                 UInt8* overflow_flag = nullptr) {
+    requires IsDataTypeDecimal<FromDataType> && IsDataTypeDecimal<ToDataType>
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_decimal.h:410]: requires () ?

File: be/src/vec/data_types/data_type_decimal.h [hunk 3]
```diff
@@ -534,9 +534,9 @@ void convert_decimal_cols(
 }
 
 template <typename FromDataType, typename ToDataType>
-std::enable_if_t<IsDataTypeDecimal<FromDataType> && IsDataTypeNumber<ToDataType>,
-                 typename ToDataType::FieldType>
-convert_from_decimal(const typename FromDataType::FieldType& value, UInt32 scale) {
+    requires IsDataTypeDecimal<FromDataType> && IsDataTypeNumber<ToDataType>
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_decimal.h:537]: requires ()?

File: be/src/vec/data_types/data_type_decimal.h [hunk 4]
```diff
@@ -578,10 +578,9 @@ convert_from_decimal(const typename FromDataType::FieldType& value, UInt32 scale
 }
 
 template <typename FromDataType, typename ToDataType>
-std::enable_if_t<IsDataTypeNumber<FromDataType> && IsDataTypeDecimal<ToDataType>,
-                 typename ToDataType::FieldType>
-convert_to_decimal(const typename FromDataType::FieldType& value, UInt32 scale,
-                   UInt8* overflow_flag) {
+    requires IsDataTypeNumber<FromDataType> && IsDataTypeDecimal<ToDataType>
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_decimal.h:581]: requires ()?

File: be/src/vec/core/decimal_comparison.h (hunk 1 above)
Comment (ucasfl) [reply be/src/vec/core/decimal_comparison.h:138]: Unneed, just need `()` when having `!`

**Comment by ucasfl** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.16 seconds
 stream load tsv:          514 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162285562 Bytes

**Review by BiteTheDDDDt — APPROVED** [review]

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**